File: 0104.Maximum_Depth_of_Binary_Tree/Solution.java
/**
 * Definition for binary tree
 * class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) {
 *      val = x;
 *      left=null;
 *      right=null;
 *     }
 * }
 */
public class Solution {
    public int maxDepth(TreeNode A) {
        if (A == null) {
            return 0;
        }
        
        int leftCounter = maxDepth(A.left);
        int rightCounter = maxDepth(A.right);
        
        return Math.max(leftCounter, rightCounter) + 1;
    }
}

File: 0104.Maximum_Depth_of_Binary_Tree/Solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
  // root - null?
  // root - single?
  // root has one child? - 1

  // Solution 1
  // brute force
  // O(N)

  return dfs(root, 0);
};

var dfs = function(node, counter) {
  if (!node) {
    return counter;
  }
  counter++;

  const leftCounter = dfs(node.left, counter);
  const rightCounter = dfs(node.right, counter);

  return Math.max(leftCounter, rightCounter);
};
